### DISCUSSION TITLE
[Feature in 5.1.23] a simple button to cycle in color palettes

### DISCUSSION THREAD

**Discussion by kookma** [TiddlyWiki/TiddlyWiki5#5271] [General]
Using the new **Show and tell** in **Discussion board** of GitHub, I would like to share some examples of the new `cycle` operator introduced in TW 5.1.23.

**Cycle in Color Palettes**

Using `cycle` filter operator it is simple to build a button to cycle in your color palettes!
You may choose few color palettes if not want all of them.

The below script cycles among all installed color palettes in your wiki

```
\define cycle-palettes()
<$set name=pals filter="[all[tiddlers+shadows]tag[$:/tags/Palette]]">
<$action-listops $tiddler="$:/palette" $field="text" $subfilter="+[cycle<pals>]"/>
</$set>
\end


<$button actions=<<cycle-palettes>> >
<$text text={{{ [[$:/palette]get[text]removeprefix[$:/palettes/]]}}}/>
</$button>
```

To give a try:

1. Go to https://tiddlywiki.com/prerelease
2. Create a new tiddler
3. Put the above script as tiddler text
4. Save

Now click the button and see how nicely the color palette changes.

**Comment by saqimtiaz** [comment]
Hi @kookma , thank you for sharing and I appreciate your enthusiasm.

I do wonder if using the Show & Tell area on Github makes sense for TiddlyWiki? The idea was to use discussions here for development related topics. If we have a lot of content like this it will lead to clutter in notifications, make life a lot harder for developers. 

Personally I think we are better off sharing such things on the Google Group considering that the audience is end users. In fact for a show and tell feature for TiddlyWiki we probably need to use TiddlyWiki so the examples can be run/seen right away!

 This is just my opinion, others might disagree :)

**Comment by kookma** [comment]
Hi @saqimtiaz - No problem.

**Comment by Jermolene** [comment]
I agree that show and tell belongs in the main group unless it's a topic that is guaranteed to only be of interest to developers.

I've removed the "Show and Tell" category for the moment. I suspect we should revisit the categories; more tightly defined categories would more clearly communicate what this forum is intended to be used for.

**Comment by kookma** [comment]
Hi Jeremy,
 No problem. Closing this category is fine because its title and GitHub description says what I did. In another thread I explained it is good to have a page a blog or something similar to explain new features with enough demo specially when a new version is released. If core has many features, but nobody knows except developers, then users cannot benefit from these features.

I am sure you have read the story of wood cutter (http://www.ece.ubc.ca/~kkleung/collection/woodcutter.htm) I think if put 20hrs a week for Tiddlywiki (I know you put much more), one half or at least one third should be for documentation, demonstrations, explaining methods for best use of Tiddlywiki and ....

Please accept my apology to have any recommendation for developers, but I see all the energy has been put for development.

Thank you for all your efforts, time and talent for giving us such a great tool free of charge!

**Reply by saqimtiaz** [reply]
>  I think if put 20hrs a week for Tiddlywiki (I know you put much more), one half or at least one third should be for documentation, demonstrations, explaining methods for best use of Tiddlywiki and ....

@kookma In my honest opinion this would be an unfair demand on a very small number of developers contributing to the core, who already ensure that every new feature is documented on tw5.com

Asking much more is likely to discourage contributions and further hurt the development aspect of the community. We already rely far too much on Jeremy for development work as a result of having so few active contributors.

> In another thread I explained it is good to have a page a blog or something similar to explain new features with enough demo specially when a new version is released. If core has many features, but nobody knows except developers, then users cannot benefit from these features.

I really think this is an area where the experienced members of the community can step up and fill in the gaps. They do not need to be developers, this example of yours illustrates that. Furthermore the blog need not, and in fact **should not**, be set up by Jeremy. This should be a community initiated venture, and if the content is good it will definitely get highlighted on tiddlywiki.com

If Jeremy was to set up an official blog, then we would have issues like having to review and ensure the quality of the information presented, which would slow down the process just as it does for the official documentation. Especially in the starting phases of a project this is the last thing that will help.

**Reply by kookma** [reply]
> I really think this is an area where the experienced members of the community can step up and fill in the gaps. They do not need to be developers, this example of yours illustrates that. Furthermore the blog need not, and in fact **should not**, be set up by Jeremy. This should be a community initiated venture, and if the content is good it will definitely get highlighted on tiddlywiki.com
> 
I know there are interested people how like to help in this regards.

> If Jeremy was to set up an official blog, then we would have issues like having to review and ensure the quality of the information presented, which would slow down the process just as it does for the official documentation. Especially in the starting phases of a project this is the last thing that will help.

@saqimtiaz - I think there is a third solution and it is an official blog, site, page, ... which is part of official Tiddlywiki.com but lets/accepts small articles, tips, solutions, ... from other users. There is no need for peer-review, they will be published by interested users with their name, other can comments, BUT time to time we see contributions by developers, super users, specially Jeremy! This way as long as Tiddlywiki.com is alive that resource is alive.

My personal experiences on Tiddlywiki shows individual efforts cannot be continued for long time, we need contribution for the whole community.

Thank you

**Reply by saqimtiaz** [reply]
> I think there is a third solution and it is an official blog, site, page, ... which is part of official Tiddlywiki.com but lets/accepts small articles, tips, solutions, ... from other users. There is no need for peer-review, they will be published by interested users with their name, other can comments, BUT time to time we see contributions by developers, super users, specially Jeremy! This way as long as Tiddlywiki.com is alive that resource is alive.

Anything "official" needs proper overview and quality control, and will therefore both be time consuming to get off the ground, will need time and input from Jeremy, and will be slow to work with just as is the case for tw5.com documentation. Honestly there is no way around that. 

> My personal experiences on Tiddlywiki shows individual efforts cannot be continued for long time, we need contribution for the whole community.

We had community initiatives that worked well for a considerable time during the peak days of TiddlyWiki Classic. The key was that it was a community effort and not by an individual.

**Reply by saqimtiaz** [reply]
PS: Where it might make sense to ask Jeremy for assistance is for just the **hosting** for a TiddlyWiki meant to be used for a community blog/wiki for show and tell.

**Comment by kookma** [comment]
@saqimtiaz - Thank you for your comments.
I think there is a misunderstanding. I don't ask Jeremy or other developers to put much more time for development or documentation. What I said and what I meant is: I _know_ you put a lot of time to keep Tiddlywiki up-to-date.

What I said is a recommendation: Having a lot of great features but few people know do not make the Product to be a successful one! It is a general point and not specific to Tiddlywiki.

I do see there are few developers and among them Jeremy, Simon, and you put a lot of time and energy to develop Tiddlywiki. I do always  appreciate this! I cannot expect you to do more, as all you, do your best!

BUT, what I say is WE need to have good demonstration, documentation for users. WE should help users to know how effectively they can employ Tiddlywiki in their personal/professional work. I like to see 1M stars for Tiddlywiki on GitHub. So, I am in your side!

I am not good at English, may be my composition caused that misunderstanding!

**Reply by saqimtiaz** [reply]
@kookma no worries. I think we have the same goals. Allocating resources is always different.

It would be great to see more community initiative in the direction of non-official documentation. Maybe a show and tell blog, hosted on a TiddlyWiki so examples can actually work? Curated by a small group of community members and everyone can submit posts for consideration?

I think its important to do it as a community initiative at first rather than trying to make it official, as that will allow far greater freedom and allow things to move much faster.

**Comment by AnthonyMuscio** [comment]
Mohammad,

Not withstanding where this thread has gone, could I suggest you take this pallet switching example and convert it to a page control button?

You could then actually submit it as a button to be included in future releases and it will appear under the tools tab in the sidebar. It is a lot of functionality, manipulates existing features, for only a few bytes And looks like a "missing feature" so I imagine it would be acceptable as a core button.

Regards
Tony

**Reply by kookma** [reply]
Hi Tony
 There is very nice working example, Notebook theme and plugin by @NicolasPetton . Look at the page controls, the last button is what you want. It simple to have just the button in your wiki.

 https://nicolas.petton.fr/tw/notebook.html
